content/blog/elm-at-noredink/index.md

```diff
@@ -0,0 +1,445 @@
+---
+title: "Elm at NoRedInk"
+date: "2021-02-15T12:00:00.000Z"
+description: How to organize 500K lines of Elm.
+---
+
+At [NoRedInk](https://www.noredink.com) we have one of the largest Elm apps
+in the world. It serves millions of teachers and students every day and our
+frontend is exclusively written in Elm. In this post, we will explore the
+structure of our application and the patterns that we use to stay sane.
+
+One of the most common questions I get about Elm is "But does it scale? And
+if so, how". Well, I hope that this post can become a comprehensive answer
+to that question.
+
+## Some stats first
+
+We have a glorious monorepo which contains all the services that power our
+infrastructure. Our main application (aka the `monolith`) is written in
+Rails and has a `ui` folder with all of our Elm code.
+
+```bash
+$ find monolith/ui/src -name "*.elm" | xargs cat | wc -l
+275082
+
+$ find monolith/ui/tests -name "*.elm" | xargs cat | wc -l
+209107
+```
+
+As you can see, we write loads of tests. As much as the Elm compiler gives
+you that wonderful bliss of refactoring, we still want to make sure that
+the code that we deploy is as correct as it can be.
+
+In the somewhat famous article ["What to know before debating type
+systems"](https://cdsmith.wordpress.com/2011/01/09/an-old-article-i-wrote/),
+two approaches are given which can help us writing correct programs:
+
+- **Testing**: establishes upper bounds on correctness
+- **Proof**: establishes lower bounds on correctness
+
+In our case, static types are _lightweight_ proofs:
+
+> An important point here is that static typing does not preclude proving
+> correctness in the traditional way, nor testing the program. It is a
+> technique to handle those cases in which testing might be guaranteed to
+> succeed so they don’t need testing; and similarly, to provide a basis
+> from which the effort of manual proof can be saved for those truly
+> challenging areas in which it is necessary.
+
+## Rails conventions
+
+Following the Rails architecture, each resource is managed by their own
+controller. So when the teacher wants to manage their classes:
+
+- they visit the `/teach/classes` URL
+- that route is managed by the `Teach::ClassesController` controller
+- that controller will fetch stuff from the database, clean it up and load a view
+
+That view looks like this:
+
+```haml
+- content_for :title, "Manage Classes | NoRedInk"
+
+- content_for(:javascript) do
+  = javascript_include_tag 'webpack/teach/classes'
+
+= elm_mount(@page_data, 'teach-classes-elm')
```

Comment:
🎨 maybe rename `@page_data` to `@page_flags`? might help readers familiar with that term understand what's going on sooner.

Reply:
Yup, thank.